libraries/node-core-library/src/LockFile.ts

```diff
@@ -0,0 +1,286 @@
+ */
+export function getProcessStartTime(pid: number): string | undefined {
+  let args: string[];
+  if (process.platform === 'darwin') {
+    args = [`-p ${pid.toString()}`, '-o lstart'];
+  } else if (process.platform === 'linux') {
+    args = ['-p', pid.toString(), '-o', 'lstart'];
+  } else {
+    throw new Error(`Unsupported system: ${process.platform}`);
+  }
+
+  const psResult: string = child_process.spawnSync('ps', args).stdout.toString();
+
+  // there was an error executing psresult
+  if (!psResult) {
+    throw new Error(`Unexpected output from "ps" command`);
+  }
+
+  const psSplit: string[] = psResult.split('\n');
+
+  // successfuly able to run "ps", but no process was found
+  if (psSplit[1] === '') {
+    return undefined;
+  }
+
+  if (psSplit[1]) {
+    const trimmed: string = psSplit[1].trim();
+    if (trimmed.length > 10) {
+      return trimmed;
+    }
+  }
+
+  throw new Error(`Unexpected output from the "ps" command`);
+}
+
+/**
+ * A helper utility for working with file-based locks.
+ * This class should only be used for locking resources across processes,
+ * but should not be used for attempting to lock a resource in the same process.
+ * @public
+ */
+export class LockFile {
+  private static _getStartTime: (pid: number) => string | undefined = getProcessStartTime;
+
+  /**
+   * Attempts to create a lockfile with the given filePath.
+   * If successful, returns a LockFile instance.
+   * If unable to get a lock, returns undefined.
+   * @param filePath - the filePath of the lockfile.
+   */
+  public static tryAcquire(filePath: string): LockFile | undefined {
+    // resolve the filepath in case the cwd is changed during process execution
+    filePath = path.resolve(filePath);
+    if (process.platform === 'win32') {
+      return LockFile._tryAcquireWindows(filePath);
+    } else if (process.platform === 'linux' || process.platform === 'darwin') {
+      return LockFile._tryAcquireMacOrLinux(filePath);
+    }
+    throw new Error(`Unable to create lock file for platform: "${process.platform}"`);
+  }
+
+  /**
+   * Attempts to acquire the lock on a Linux or OSX machine
+   */
+  private static _tryAcquireMacOrLinux(absoluteFilePath: string): LockFile | undefined {
+    let dirtyWhenAcquired: boolean = false;
+
+    // get the current process' pid
+    const pid: number = process.pid;
+    const startTime: string | undefined = LockFile._getStartTime(pid);
+
+    if (!startTime) {
+      throw new Error(`UNable to calcualte start time for current process.`);
+    }
+
+    const lockFileDir: string = path.dirname(absoluteFilePath);
+
+    // create your own pid file
+    const pidLockFilePath: string = `${absoluteFilePath}.${pid}`;
+    let lockFileDescriptor: number | undefined;
+
+    let lockFile: LockFile;
+
+    try {
+      // open in write mode since if this file exists, it cannot be from the current process
+      // we are not handling the case where multiple lockfiles are opened in the same process
```

Comment:
>/ we are not handling the case where multiple lockfiles are opened in the same process [](start = 7, length = 86)

TODO: This will malfunction if the same process tries to acquire two locks on the same file.  We should ideally maintain a dictionary of normalized acquired filenames